float32 desired
float32 velocity